File: sdks/sdkwork-mail-sdk/sdkwork-mail-sdk-kotlin/src/main/kotlin/com/sdkwork/Mail/metadata/mailProviderPackageLoader.kt
package com.sdkwork.Mail.metadata

class MailProviderPackageLoaderException(
    val code: String,
    override val message: String,
) : RuntimeException(message)

data class MailProviderPackageLoadRequest(
    val providerKey: String? = null,
    val packageIdentity: String? = null,
)

data class MailResolvedProviderPackageLoadTarget(
    val packageEntry: MailProviderPackageCatalogEntry,
)

typealias MailProviderModuleNamespace = Map<String, String>
typealias MailProviderPackageImportFn = (MailResolvedProviderPackageLoadTarget) -> MailProviderModuleNamespace
typealias MailProviderPackageLoader = (MailProviderPackageLoadRequest) -> MailProviderModuleNamespace

data class MailProviderPackageInstallRequest(
    val driverManager: Any,
    val loadRequest: MailProviderPackageLoadRequest,
)

fun resolveMailProviderPackageLoadTarget(
    request: MailProviderPackageLoadRequest,
): MailResolvedProviderPackageLoadTarget {
    val packageByProviderKey = request.providerKey?.let { MailProviderPackageCatalog.getMailProviderPackageByProviderKey(it) }
    val packageByIdentity = request.packageIdentity?.let { MailProviderPackageCatalog.getMailProviderPackageByPackageIdentity(it) }

    if (packageByProviderKey != null
        && packageByIdentity != null
        && packageByProviderKey.packageIdentity != packageByIdentity.packageIdentity
    ) {
        throw MailProviderPackageLoaderException(
            code = "provider_package_identity_mismatch",
            message = "providerKey and packageIdentity must resolve to the same provider package boundary.",
        )
    }

    val resolvedPackage = packageByProviderKey ?: packageByIdentity
        ?: throw MailProviderPackageLoaderException(
            code = "provider_package_not_found",
            message = "No official provider package matches the requested provider boundary.",
        )

    return MailResolvedProviderPackageLoadTarget(packageEntry = resolvedPackage)
}

fun createMailProviderPackageLoader(
    importPackage: MailProviderPackageImportFn,
): MailProviderPackageLoader = { request ->
    loadMailProviderModule(request, importPackage)
}

fun loadMailProviderModule(
    request: MailProviderPackageLoadRequest,
    importPackage: MailProviderPackageImportFn,
): MailProviderModuleNamespace {
    val target = resolveMailProviderPackageLoadTarget(request)

    return try {
        val namespace = importPackage(target)
        if (namespace.isEmpty()) {
            throw MailProviderPackageLoaderException(
                code = "provider_module_export_missing",
                message = "Reserved provider package loader scaffold requires an executable provider module namespace.",
            )
        }

        namespace
    } catch (error: MailProviderPackageLoaderException) {
        throw error
    } catch (error: RuntimeException) {
        throw MailProviderPackageLoaderException(
            code = "provider_package_load_failed",
            message = "Reserved provider package loader scaffold could not load ${target.packageEntry.packageIdentity}: ${error.message}",
        )
    }
}

fun installMailProviderPackage(
    request: MailProviderPackageInstallRequest,
    importPackage: MailProviderPackageImportFn,
) {
    loadMailProviderModule(request.loadRequest, importPackage)

    throw MailProviderPackageLoaderException(
        code = "provider_package_load_failed",
        message = "Reserved provider package installer scaffold cannot register provider modules until a verified runtime bridge lands.",
    )
}

fun installMailProviderPackages(
    requests: Iterable<MailProviderPackageInstallRequest>,
    importPackage: MailProviderPackageImportFn,
) {
    val materializedRequests = requests.toList()
    materializedRequests.forEach { request ->
        loadMailProviderModule(request.loadRequest, importPackage)
    }

    if (materializedRequests.isNotEmpty()) {
        throw MailProviderPackageLoaderException(
            code = "provider_package_load_failed",
            message = "Reserved provider package installer scaffold cannot register provider modules until a verified runtime bridge lands.",
        )
    }
}

File: sdks/sdkwork-mail-sdk/sdkwork-mail-sdk-swift/Sources/MailSdk/mailProviderPackageLoader.swift
public struct MailProviderPackageLoaderException: Error {
    public let code: String
    public let message: String

    public init(code: String, message: String) {
        self.code = code
        self.message = message
    }
}

public struct MailProviderPackageLoadRequest {
    public let providerKey: String?
    public let packageIdentity: String?

    public init(providerKey: String? = nil, packageIdentity: String? = nil) {
        self.providerKey = providerKey
        self.packageIdentity = packageIdentity
    }
}

public struct MailResolvedProviderPackageLoadTarget {
    public let packageEntry: MailProviderPackageCatalogEntry

    public init(packageEntry: MailProviderPackageCatalogEntry) {
        self.packageEntry = packageEntry
    }
}

public typealias MailProviderModuleNamespace = [String: String]
public typealias MailProviderPackageImportFn = (MailResolvedProviderPackageLoadTarget) throws -> MailProviderModuleNamespace
public typealias MailProviderPackageLoader = (MailProviderPackageLoadRequest) throws -> MailProviderModuleNamespace

public struct MailProviderPackageInstallRequest {
    public let driverManager: Any
    public let loadRequest: MailProviderPackageLoadRequest

    public init(driverManager: Any, loadRequest: MailProviderPackageLoadRequest) {
        self.driverManager = driverManager
        self.loadRequest = loadRequest
    }
}

public func resolveMailProviderPackageLoadTarget(
    _ request: MailProviderPackageLoadRequest
) throws -> MailResolvedProviderPackageLoadTarget {
    let packageByProviderKey = request.providerKey.flatMap(MailProviderPackageCatalog.getMailProviderPackageByProviderKey)
    let packageByIdentity = request.packageIdentity.flatMap(MailProviderPackageCatalog.getMailProviderPackageByPackageIdentity)

    if let providerKeyEntry = packageByProviderKey,
       let packageIdentityEntry = packageByIdentity,
       providerKeyEntry.packageIdentity != packageIdentityEntry.packageIdentity {
        throw MailProviderPackageLoaderException(
            code: "provider_package_identity_mismatch",
            message: "providerKey and packageIdentity must resolve to the same provider package boundary."
        )
    }

    guard let resolvedPackage = packageByProviderKey ?? packageByIdentity else {
        throw MailProviderPackageLoaderException(
            code: "provider_package_not_found",
            message: "No official provider package matches the requested provider boundary."
        )
    }

    return MailResolvedProviderPackageLoadTarget(packageEntry: resolvedPackage)
}

public func createMailProviderPackageLoader(
    importPackage: @escaping MailProviderPackageImportFn
) -> MailProviderPackageLoader {
    return { request in
        try loadMailProviderModule(request, importPackage: importPackage)
    }
}

public func loadMailProviderModule(
    _ request: MailProviderPackageLoadRequest,
    importPackage: MailProviderPackageImportFn
) throws -> MailProviderModuleNamespace {
    let target = try resolveMailProviderPackageLoadTarget(request)

    do {
        let namespace = try importPackage(target)
        if namespace.isEmpty {
            throw MailProviderPackageLoaderException(
                code: "provider_module_export_missing",
                message: "Reserved provider package loader scaffold requires an executable provider module namespace."
            )
        }

        return namespace
    } catch let error as MailProviderPackageLoaderException {
        throw error
    } catch {
        throw MailProviderPackageLoaderException(
            code: "provider_package_load_failed",
            message: "Reserved provider package loader scaffold could not load \(target.packageEntry.packageIdentity): \(error)"
        )
    }
}

public func installMailProviderPackage(
    _ request: MailProviderPackageInstallRequest,
    importPackage: MailProviderPackageImportFn
) throws {
    _ = try loadMailProviderModule(request.loadRequest, importPackage: importPackage)

    throw MailProviderPackageLoaderException(
        code: "provider_package_load_failed",
        message: "Reserved provider package installer scaffold cannot register provider modules until a verified runtime bridge lands."
    )
}

public func installMailProviderPackages(
    _ requests: [MailProviderPackageInstallRequest],
    importPackage: MailProviderPackageImportFn
) throws {
    for request in requests {
        _ = try loadMailProviderModule(request.loadRequest, importPackage: importPackage)
    }

    if !requests.isEmpty {
        throw MailProviderPackageLoaderException(
            code: "provider_package_load_failed",
            message: "Reserved provider package installer scaffold cannot register provider modules until a verified runtime bridge lands."
        )
    }
}
